Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp

```diff
@@ -0,0 +1,298 @@
+        if (!upperBound)
+            return nullptr;
+
+        // If the lower bound of any range is higher than the upper bound, the entire descriptor is invalid and must be
+        // ignored.
+        if (lowerBound->isInteger() && upperBound->isInteger() && lowerBound->resolveAsIntegerDeprecated() > upperBound->resolveAsIntegerDeprecated())
+            return nullptr;
+
+        return CSSValuePair::createNoncoalescing(lowerBound.releaseNonNull(), upperBound.releaseNonNull());
+    });
+
+    if (!range.atEnd() || !rangeList || !rangeList->length())
+        return nullptr;
+    return rangeList;
+}
+
+RefPtr<CSSValue> consumeCounterStylePad(CSSParserTokenRange& range, const CSSParserContext& context)
+{
+    // <'pad'> = <integer [0,∞]> && <symbol>
+    // https://drafts.csswg.org/css-counter-styles-3/#counter-style-pad
+
+    RefPtr<CSSPrimitiveValue> integer;
+    RefPtr<CSSValue> symbol;
+    while (!integer || !symbol) {
+        if (!integer) {
+            integer = consumeNonNegativeInteger(range);
+            if (integer)
+                continue;
+        }
+        if (!symbol) {
+            symbol = consumeCounterStyleSymbol(range, context);
+            if (symbol)
+                continue;
+        }
+        return nullptr;
+    }
+    if (!range.atEnd())
+        return nullptr;
+    return CSSValueList::createSpaceSeparated(integer.releaseNonNull(), symbol.releaseNonNull());
+}
+
+RefPtr<CSSValue> consumeCounterStyleSymbols(CSSParserTokenRange& range, const CSSParserContext& context)
+{
+    // <'symbols'> = <symbol>+
+    // https://drafts.csswg.org/css-counter-styles-3/#descdef-counter-style-symbols
+
+    CSSValueListBuilder symbols;
+    while (!range.atEnd()) {
+        auto symbol = consumeCounterStyleSymbol(range, context);
+        if (!symbol)
+            return nullptr;
+        symbols.append(symbol.releaseNonNull());
+    }
+    if (symbols.isEmpty())
+        return nullptr;
+    return CSSValueList::createSpaceSeparated(WTFMove(symbols));
+}
+
+RefPtr<CSSValue> consumeCounterStyleAdditiveSymbols(CSSParserTokenRange& range, const CSSParserContext& context)
+{
+    // <'additive-symbols'> = [ <integer [0,∞]> && <symbol> ]#
+    // https://drafts.csswg.org/css-counter-styles-3/#descdef-counter-style-additive-symbols
+
+    std::optional<int> lastWeight;
+    auto values = consumeCommaSeparatedListWithoutSingleValueOptimization(range, [&lastWeight](auto& range, auto& context) -> RefPtr<CSSValue> {
+        auto integer = consumeNonNegativeInteger(range);
+        auto symbol = consumeCounterStyleSymbol(range, context);
+        if (!integer) {
+            if (!symbol)
+                return nullptr;
+            integer = consumeNonNegativeInteger(range);
+            if (!integer)
+                return nullptr;
+        }
+        if (!symbol)
+            return nullptr;
+
+        // Additive tuples must be specified in order of strictly descending weight.
```

Comment:
Do we need test cases for how this works with `calc`? Seems subtle.

Reply:
I am not yet sure how this is supposed to work. Since this is an @counter-style rule descriptor, the usual time for calc() resolution does not apply and https://drafts.csswg.org/css-counter-styles-3/ doesn't define what to do.

I have filed https://github.com/w3c/csswg-drafts/issues/10853 to start the conversation with editors about this class of problem.

For now, this is just moved code, so nothing is needed in this PR.

Reply:
(adding those "Deprecated" named functions was to make it really easy to find all the places, like this, that we need to figure out).